receiver/receiverhelper/receiver.go

```diff
@@ -0,0 +1,347 @@
+		defaultCollectionInterval: time.Minute,
+		nextConsumer:              nextConsumer,
+		done:                      make(chan struct{}),
+	}
+
+	for _, op := range options {
+		op(mr)
+	}
+
+	// wrap the start function with a call to initialize scrapers
+	// and start scraping
+	start := mr.start
+	mr.start = func(ctx context.Context, host component.Host) error {
+		if start != nil {
+			if err := start(ctx, host); err != nil {
+				return err
+			}
+		}
+
+		if err := mr.initializeScrapers(ctx); err != nil {
+			return err
+		}
+
+		mr.startScraping()
+		return nil
+	}
+
+	// wrap the shutdown function with a call to close scrapers
+	// and stop scraping
+	shutdown := mr.shutdown
+	mr.shutdown = func(ctx context.Context) error {
+		mr.stopScraping()
+
+		var errors []error
+
+		if err := mr.closeScrapers(ctx); err != nil {
+			errors = append(errors, err)
+		}
+
+		if shutdown != nil {
+			if err := shutdown(ctx); err != nil {
+				errors = append(errors, err)
+			}
+		}
+
+		return componenterror.CombineErrors(errors)
+	}
+
+	return mr, nil
+}
+
+// initializeScrapers initializes all the scrapers
+func (mr *metricsReceiver) initializeScrapers(ctx context.Context) error {
+	for _, scraper := range mr.allBaseScrapers() {
+		if scraper.initialize == nil {
+			continue
+		}
+
+		if err := scraper.initialize(ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// startScraping initiates a ticker that calls Scrape based on the configured
+// collection interval.
+func (mr *metricsReceiver) startScraping() {
+	// TODO: use one ticker for each set of scrapers that have the same collection interval.
+
+	mr.startScrapingMetrics()
+	mr.startScrapingResourceMetrics()
+}
+
+func (mr *metricsReceiver) startScrapingMetrics() {
+	for i := 0; i < len(mr.metricScrapers); i++ {
+		scraper := mr.metricScrapers[i]
+		go func() {
+			collectionInterval := mr.defaultCollectionInterval
+			if scraper.cfg.CollectionInterval() != 0 {
+				collectionInterval = scraper.cfg.CollectionInterval()
+			}
+
+			ticker := time.NewTicker(collectionInterval)
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ticker.C:
+					mr.scrapeMetricsAndReport(context.Background(), scraper)
+				case <-mr.done:
+					return
+				}
+			}
+		}()
+	}
+}
+
+// scrapeMetricsAndReport calls the Scrape function of the provided Metrics Scraper,
+// records observability information, and passes the scraped metrics to the next
+// component.
+func (mr *metricsReceiver) scrapeMetricsAndReport(ctx context.Context, ms *metricsScraper) {
+	// TODO: Add observability metrics support
+	metrics, err := ms.scrape(ctx)
+	if err != nil {
+		return
+	}
+
+	mr.nextConsumer.ConsumeMetrics(ctx, metricSliceToMetricData(metrics))
+}
+
+func metricSliceToMetricData(metrics pdata.MetricSlice) pdata.Metrics {
+	rms := pdata.NewResourceMetricsSlice()
+	rms.Resize(1)
+	rm := rms.At(0)
+	ilms := rm.InstrumentationLibraryMetrics()
+	ilms.Resize(1)
+	ilm := ilms.At(0)
```

Comment:
For metrics that we "scrape", would we want to fill in Instrumentation Library data btw?

Reply:
It's possible, see https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1099